scipy/differentiate/tests/test_differentiate.py

```diff
@@ -473,177 +472,190 @@ def test_iv(self):
-        return [[np.sin(phi) * np.cos(th), r * np.cos(phi) * np.cos(th),
-                 -r * np.sin(phi) * np.sin(th)],
-                [np.sin(phi) * np.sin(th), r * np.cos(phi) * np.sin(th),
-                 r * np.sin(phi) * np.cos(th)],
-                [np.cos(phi), -r * np.sin(phi), np.zeros_like(r)]]
+        return [[xp.sin(phi) * xp.cos(th), r * xp.cos(phi) * xp.cos(th),
+                 -r * xp.sin(phi) * xp.sin(th)],
+                [xp.sin(phi) * xp.sin(th), r * xp.cos(phi) * xp.sin(th),
+                 r * xp.sin(phi) * xp.cos(th)],
+                [xp.cos(phi), -r * xp.sin(phi), xp.zeros_like(r)]]
 
     f3.mn = 3, 3  # type: ignore[attr-defined]
     f3.ref = df3  # type: ignore[attr-defined]
 
-    def f4(x):
+    def f4(x, xp):
         x1, x2, x3 = x
-        return [x1, 5 * x3, 4 * x2 ** 2 - 2 * x3, x3 * np.sin(x1)]
+        return [x1, 5 * x3, 4 * x2 ** 2 - 2 * x3, x3 * xp.sin(x1)]
 
-    def df4(x):
+    def df4(x, xp):
         x1, x2, x3 = x
-        one = np.ones_like(x1)
+        one = xp.ones_like(x1)
         return [[one, 0 * one, 0 * one],
                 [0 * one, 0 * one, 5 * one],
                 [0 * one, 8 * x2, -2 * one],
-                [x3 * np.cos(x1), 0 * one, np.sin(x1)]]
+                [x3 * xp.cos(x1), 0 * one, xp.sin(x1)]]
 
     f4.mn = 3, 4  # type: ignore[attr-defined]
     f4.ref = df4  # type: ignore[attr-defined]
 
-    def f5(x):
+    def f5(x, xp):
         x1, x2, x3 = x
-        return [5 * x2, 4 * x1 ** 2 - 2 * np.sin(x2 * x3), x2 * x3]
+        return [5 * x2, 4 * x1 ** 2 - 2 * xp.sin(x2 * x3), x2 * x3]
 
-    def df5(x):
+    def df5(x, xp):
         x1, x2, x3 = x
-        one = np.ones_like(x1)
+        one = xp.ones_like(x1)
         return [[0 * one, 5 * one, 0 * one],
-                [8 * x1, -2 * x3 * np.cos(x2 * x3), -2 * x2 * np.cos(x2 * x3)],
+                [8 * x1, -2 * x3 * xp.cos(x2 * x3), -2 * x2 * xp.cos(x2 * x3)],
                 [0 * one, x3, x2]]
 
     f5.mn = 3, 3  # type: ignore[attr-defined]
     f5.ref = df5  # type: ignore[attr-defined]
 
-    rosen = optimize.rosen
+    def rosen(x, xp): return optimize.rosen(x)
     rosen.mn = 5, 1  # type: ignore[attr-defined]
-    rosen.ref = optimize.rosen_der  # type: ignore[attr-defined]
+    def ref(x, xp): return optimize.rosen_der(x)
+    rosen.ref = ref  # type: ignore[attr-defined]
 
     @pytest.mark.parametrize('size', [(), (6,), (2, 3)])
     @pytest.mark.parametrize('func', [f1, f2, f3, f4, f5, rosen])
-    def test_examples(self, size, func):
+    def test_examples(self, size, func, xp):
         rng = np.random.default_rng(458912319542)
         m, n = func.mn
-        x = rng.random(size=(m,) + size)
-        res = jacobian(func, x)
-        ref = func.ref(x)
-        np.testing.assert_allclose(res.df, ref, atol=1e-10)
+        x = xp.asarray(rng.random(size=(m,) + size), dtype=xp.asarray(1.).dtype)
+        res = jacobian(lambda x: func(x, xp), x)
+        ref = xp.asarray(func.ref(x, xp))
```

Comment:
I got a little stuck here because we end up with nested lists of tensors depending on `size` which throws an error when calling `xp.asarray` on. I wanted to check in here before going down a nested `xp.stack` route

Reply:
Actually I should just compute the reference values with numpy then call `xp.asarray`